### PR TITLE
Batching: DataStore + related components

### PR DESCRIPTION
**Pull request by charlieyou**
Part of the Batching MVP Implementation

DataStore is the API that is used for communication between the ModelThread and PredictFrontend when batching is enabled.
It abstracts away the Queue, HashMap, and Codecs. The Queue and HashMap have implementations for their associated Redis data structures.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: docker/wsgi.py [hunk 1]
```diff
@@ -88,6 +88,12 @@ def read_models_from_file():
 models = read_models_from_file()
 # Parse the arguments
 arguments = ArgParser.extract_args(args)
+
+# Start Redis if batching is enabled
+if arguments.batching == 'true':
+    assert arguments.redis_conf
+    subprocess.call(["redis-server", arguments.redis_conf])
```
Comment (vdantu) [inline docker/wsgi.py]: What happens if this fails?

File: mms/data_store/data_store.py [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    """
+    DataStore abstracts the queue, hashmap, and associated codecs
+    """
+
+    def __init__(self, name_prefix, config):
+        self.hashmap = RedisHashMap(name_prefix, config)
+        self.queue = RedisQueue(name_prefix, config)
+
+    def pop_batch(self, model_name, length, data_type):
+        codec = self._get_codec(data_type)
```
Comment (vdantu) [inline mms/data_store/data_store.py]: Maybe a good idea to check `codec` == none here and assert or just assert/raise exception in get_codec.

File: mms/data_store/data_store.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+class DataStore(object):
+    """
+    DataStore abstracts the queue, hashmap, and associated codecs
+    """
+
+    def __init__(self, name_prefix, config):
+        self.hashmap = RedisHashMap(name_prefix, config)
+        self.queue = RedisQueue(name_prefix, config)
+
+    def pop_batch(self, model_name, length, data_type):
```
Comment (vdantu) [inline mms/data_store/data_store.py]: Add doc string for each string in the class, mentioning its purpose.

File: mms/data_store/hashmap.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
+# Copyright 2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (vdantu) [inline mms/data_store/hashmap.py]: copyrights probably should change to 2018.  Just a nitpick

File: setup.py [hunk 5]
```diff
@@ -25,7 +25,7 @@
 # $ twine upload dist/*
 requirements = ['Flask', 'Pillow', 'requests', 'flask-cors',
                 'psutil', 'jsonschema', 'onnx-mxnet>=0.4.2', 'boto3', 'importlib2',
-                'fasteners']
+                'fasteners', 'redis', 'numpy']
```
Comment (vdantu) [inline setup.py]: Where is numpy used?

File: mms/data_store/hashmap.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+        self.name_prefix = name_prefix
+        try:
+            self.redis = redis.StrictRedis(**config)
+        except Exception as e:
+            raise Exception("Failed to connect to Redis: %s" % e)
+
+    def put(self, key, data):
+        self.redis.lpush(self.name_prefix + key, data)
+
+    def get(self, key, timeout=60):
```
Comment (vdantu) [inline mms/data_store/hashmap.py]: 60 seconds blocked call? Isn't that too large?

File: mms/data_store/queue.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+# Licensed under the Apache License, Version 2.0 (the "License").
+# You may not use this file except in compliance with the License.
+# A copy of the License is located at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# or in the "license" file accompanying this file. This file is distributed
+# on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+# express or implied. See the License for the specific language governing
+# permissions and limitations under the License.
+
+"""
```
Comment (vdantu) [inline mms/data_store/queue.py]: Add docstrings to all the methods

File: mms/redis_conf.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+"""
+This module parses a Redis configuration file to ensure the same parameters
+used to initialize Redis are used by the Redis object to access Redis.
+"""
+
+CONFIG_MAP = {"bind": "host", "port": "port", "requirepass": "password"}
+
+
+class RedisConfParser(object):
```
Comment (vdantu) [inline mms/redis_conf.py]: Where is this being used?

File: mms/serving_frontend.py [hunk 9]
```diff
@@ -40,9 +43,18 @@ def __init__(self, app_name):
         app_name : str
             App name to initialize request handler.
         """
+        def gen_prefix(prefix, length=5, chars=string.ascii_letters+string.digits):
+            return prefix + ''.join(random.choice(chars) for _ in range(length)) + '_'
```
Comment (vdantu) [inline mms/serving_frontend.py]: Wouldn't we need to add a `seed` to random?

File: docker/redis.conf [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+timeout 0
+tcp-keepalive 300
+
+################################# GENERAL #####################################
+daemonize yes
+supervised no
+pidfile redis.pid
+loglevel notice
+logfile redis.log
+always-show-logo yes
```
[the remark this answers is not in the document]
Comment (charlieyou) [reply docker/redis.conf]: done

File: docker/wsgi.py (hunk 1 above)
Comment (charlieyou) [reply docker/wsgi.py]: done

Comment (vdantu) [reply docker/wsgi.py]: Can we move this outside of wsgi, probably to setup? This is called once per worker right?

File: mms/redis_conf.py (hunk 8 above)
Comment (charlieyou) [reply mms/redis_conf.py]: part of another pr, can move it to this one if necessary

File: mms/data_store/hashmap.py (hunk 6 above)
Comment (charlieyou) [reply mms/data_store/hashmap.py]: that is just the default, can be changed by the user via cli (in a later pr)

File: mms/data_store/queue.py (hunk 7 above)
Comment (charlieyou) [reply mms/data_store/queue.py]: done

File: mms/data_store/data_store.py (hunk 3 above)
Comment (charlieyou) [reply mms/data_store/data_store.py]: done

**Pull request merged**